Refactor OptimizeDateOrDateTimeConverterWithPreimageVisitor

ZhiguoZh opened the pull request:
The generateOptimizedDateFilter function is refactored to enhance the code readability. And this commit also fixes the duplicate creations of ColumnNode.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
As the review comments in #52091 suggested, the implementation of date filter generation is not neat enough due to the duplicate code. In this refactoring work, we try to have `createFunctionNode` create a resolved function node with the given name and a variable number of arguments, so that the readability could be refined. And at the same time, we found that the original implementation may introduce potential issues by creating `ColumnNode` repeatedly with the same column source, leading to different `id` for the same column node. We've fixed this issue together with the `EXPLAIN QUERY TREE` tests.

The performance tests of SSB has shown that this code change would not lead to performance loss.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

ZhiguoZh committed:
Refactor OptimizeDateOrDateTimeConverterWithPreimageVisitor

The generateOptimizedDateFilter function is refactored to enhance
the code readability. And this commit also fixes the duplicate
creations of ColumnNode.

vdimir reviewed (approved):
Thanks for taking the time to improve the project!

robot-ch-test-poll commented:
<!-- automatic status comment for PR #60875 from ZhiguoZh/ClickHouse:20240305-analyzer-refactor -->
*This is an automated comment for commit c4009a16f90920e3ce59b1376d5bc7529aa8d450 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60875/c4009a16f90920e3ce59b1376d5bc7529aa8d450/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

ZhiguoZh committed:
Fix functional test analyzer_preimage_null

ColumnNodes with the same source are expected to share the exact id
after the preimage transformation.

ZhiguoZh commented:
Actually I'd appreciate your patience and review as this refactoring work should have been done earlier! BTW, I observed a few failures in the tests, but it seems that they are not related to this change and are caused by the connection issues and timeout errors. Could you share with me some hint if the fixes are needed? Thanks!

vdimir commented:
CI failures https://github.com/ClickHouse/ClickHouse/pull/60875/commits/ba8a2eb10d1a2a2beb0fd72b5191c3f4b2fdbd34

> AST fuzzer (asan) — Received signal 11 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60875/ba8a2eb10d1a2a2beb0fd72b5191c3f4b2fdbd34/ast_fuzzer__asan_.html)

Looks like bug in range dictionary, repo:

<details>
<summary>Repro</summary>

```sql
DROP DICTIONARY IF EXISTS range_hashed_dictionary;
DROP TABLE IF EXISTS dictionary_source_table;
CREATE TABLE dictionary_source_table
(
    key UInt64,
    start UInt64,
    end UInt64,
    value String
) Engine = TinyLog;

INSERT INTO dictionary_source_table values (0, 0, 5, 'Value');

DROP DICTIONARY IF EXISTS range_hashed_dictionary;
CREATE DICTIONARY range_hashed_dictionary
(
    key UInt64,
    start UInt64,
    end UInt64,
    value String
)
PRIMARY KEY key
SOURCE(CLICKHOUSE(TABLE 'dictionary_source_table'))
LAYOUT(RANGE_HASHED())
RANGE(MIN start MAX end)
LIFETIME(0);

SET allow_experimental_analyzer = 1;
SELECT  toLowCardinality(4), dictGetOrDefault('range_hashed_dictionary', 'value', 1, toUInt64(materialize(toLowCardinality(6))), toFixedString(toFixedString(toFixedString(toFixedString('DefaultValue', 12), 12), 12), 12));
```
</details>

> Stateless tests (release, analyzer, s3, DatabaseReplicated) [2/4] — fail: 1, passed: 1552, skipped: 34 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60875/ba8a2eb10d1a2a2beb0fd72b5191c3f4b2fdbd34/stateless_tests__release__analyzer__s3__databasereplicated__[2_4].html)

`02724_show_indexes` is [flaky](
https://play.clickhouse.com/play?user=play#CldJVEgKICAgIGZhbHNlIGFzIG9ubHlfbWFzdGVyClNFTEVDVCAqCkZST00gKApTRUxFQ1QKICAgIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHJudW0sCiAgICBzdWJzdHIoY29tbWl0X3NoYSwgMSw5KSBhcyBzaGEsCiAgICBjaGVja19zdGFydF90aW1lLAogICAgZm9ybWF0UmVhZGFibGVUaW1lRGVsdGEoaW50RGl2KG5vdygpIC0gY2hlY2tfc3RhcnRfdGltZSwgMzYwMCkgKiAzNjAwKSBhcyBhZ28sCiAgICBjaGVja19uYW1lLAogICAgdGVzdF9uYW1lLAogICAgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDcgREFZCiAgICBBTkQgKHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwIE9SIG5vdCBvbmx5X21hc3RlcikKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyBMSUtFICdGJScKICAgIEFORCBjaGVja19zdGF0dXMgIT0gJ3N1Y2Nlc3MnCikKV0hFUkUgdGVzdF9uYW1lIGxpa2UgJyUwMjcyNF9zaG93X2luZGV4ZXMlJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIERFU0MsIGNoZWNrX25hbWUsIHRlc3RfbmFtZQ==) with analyzer

> Stateless tests (tsan) [4/5] — Tests are not finished, fail: 7, passed: 505, skipped: 3 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60875/ba8a2eb10d1a2a2beb0fd72b5191c3f4b2fdbd34/stateless_tests__tsan__[4_5].html)

<details>
<summary>ThreadSanitizer report</summary>

```
WARNING: ThreadSanitizer: data race (pid=772)
  Write of size 8 at 0x7b48000054c0 by thread T2046 (mutexes: write M0):
    #0 std::__1::enable_if<is_move_constructible<DB::PageCache::Mmap*>::value && is_move_assignable<DB::PageCache::Mmap*>::value, void>::type std::__1::swap[abi:v15000]<DB::PageCache::Mmap*>(DB::PageCache::Mmap*&, DB::PageCache::Mmap*&) build_docker/./contrib/llvm-project/libcxx/include/__utility/swap.h:36:7 (clickhouse+0x161a28b7) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 std::__1::vector<DB::PageCache::Mmap, std::__1::allocator<DB::PageCache::Mmap>>::__swap_out_circular_buffer(std::__1::__split_buffer<DB::PageCache::Mmap, std::__1::allocator<DB::PageCache::Mmap>&>&) build_docker/./contrib/llvm-project/libcxx/include/vector:917:5 (clickhouse+0x161a28b7)
    #2 void std::__1::vector<DB::PageCache::Mmap, std::__1::allocator<DB::PageCache::Mmap>>::__emplace_back_slow_path<unsigned long&, unsigned long&, unsigned long&, unsigned long&, void*&, bool&>(unsigned long&, unsigned long&, unsigned long&, unsigned long&, void*&, bool&) build_docker/./contrib/llvm-project/libcxx/include/vector:1581:5 (clickhouse+0x161a28b7)
    #3 DB::PageCache::Mmap& std::__1::vector<DB::PageCache::Mmap, std::__1::allocator<DB::PageCache::Mmap>>::emplace_back<unsigned long&, unsigned long&, unsigned long&, unsigned long&, void*&, bool&>(unsigned long&, unsigned long&, unsigned long&, unsigned long&, void*&, bool&) build_docker/./contrib/llvm-project/libcxx/include/vector:1600:9 (clickhouse+0x1619f38b) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #4 DB::PageCache::addMmap(std::__1::unique_lock<std::__1::mutex>&) build_docker/./src/Common/PageCache.cpp:556:11 (clickhouse+0x1619f38b)
    #5 DB::PageCache::getFreeChunk(std::__1::unique_lock<std::__1::mutex>&) build_docker/./src/Common/PageCache.cpp:511:9 (clickhouse+0x1619de70) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #6 DB::PageCache::getOrSet(wide::integer<128ul, unsigned int>, bool, bool) build_docker/./src/Common/PageCache.cpp:302:21 (clickhouse+0x1619d519) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #7 DB::CachedInMemoryReadBufferFromFile::nextImpl() build_docker/./src/IO/CachedInMemoryReadBufferFromFile.cpp:100:24 (clickhouse+0x161989a2) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #8 DB::ReadBuffer::next() build_docker/./src/IO/ReadBuffer.h:70:20 (clickhouse+0x16180ebc) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #9 DB::ReadBufferFromRemoteFSGather::readImpl() build_docker/./src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:213:32 (clickhouse+0x16180ebc)
    #10 DB::ReadBufferFromRemoteFSGather::nextImpl() build_docker/./src/Disks/IO/ReadBufferFromRemoteFSGather.cpp:185:9 (clickhouse+0x16180dc9) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #11 DB::ReadBuffer::next() build_docker/./src/IO/ReadBuffer.h:70:20 (clickhouse+0x15fa11eb) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #12 DB::ThreadPoolRemoteFSReader::execute(DB::IAsynchronousReader::Request, bool) build_docker/./src/Disks/IO/ThreadPoolRemoteFSReader.cpp:147:25 (clickhouse+0x15fa11eb)
    #13 DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0::operator()() const build_docker/./src/Disks/IO/ThreadPoolRemoteFSReader.cpp:110:46 (clickhouse+0x15fa29f3) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #14 std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda'()::operator()() build_docker/./src/Common/threadPoolCallbackRunner.h:42:20 (clickhouse+0x15fa29f3)
    #15 decltype(std::declval<DB::IAsynchronousReader::Result>()()) std::__1::__invoke[abi:v15000]<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda'()&>(DB::IAsynchronousReader::Result&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x15fa29f3)
    #16 std::__1::__packaged_task_func<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda'(), std::__1::allocator<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda'()>, DB::IAsynchronousReader::Result ()>::operator()() build_docker/./contrib/llvm-project/libcxx/include/future:1694:12 (clickhouse+0x15fa29f3)
    #17 std::__1::__packaged_task_function<DB::IAsynchronousReader::Result ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/future:1876:12 (clickhouse+0x15f9e860) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #18 std::__1::packaged_task<DB::IAsynchronousReader::Result ()>::operator()() build_docker/./contrib/llvm-project/libcxx/include/future:1953:24 (clickhouse+0x15f9e860)
    #19 std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda0'()::operator()() const build_docker/./src/Common/threadPoolCallbackRunner.h:50:70 (clickhouse+0x15fa2bcc) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #20 decltype(std::declval<DB::IAsynchronousReader::Result>()()) std::__1::__invoke[abi:v15000]<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda0'()&>(DB::IAsynchronousReader::Result&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x15fa2bcc)
    #21 void std::__1::__invoke_void_return_wrapper<void, true>::__call<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda0'()&>(std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda0'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x15fa2bcc)
    #22 std::__1::__function::__default_alloc_func<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda0'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x15fa2bcc)
    #23 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<std::__1::function<std::__1::future<DB::IAsynchronousReader::Result> (DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)> DB::threadPoolCallbackRunner<DB::IAsynchronousReader::Result, DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::operator()(DB::ThreadPoolRemoteFSReader::submit(DB::IAsynchronousReader::Request)::$_0&&, Priority)::'lambda0'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x15fa2bcc)
    #24 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xf209b2e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #25 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xf209b2e)
    #26 ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xf209b2e)
    #27 void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xf20defc) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #28 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20defc)
    #29 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0xf20defc)
    #30 decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0xf20defc)
    #31 ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:223:13 (clickhouse+0xf20defc)
    #32 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20de61) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #33 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0xf20de61)
    #34 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0xf20de61)
    #35 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0xf20de61)
    #36 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xf205e0e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #37 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xf205e0e)
    #38 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xf205e0e)
    #39 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xf20bcd1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #40 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20bcd1)
    #41 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0xf20bcd1)
    #42 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0xf20bcd1)

  Previous read of size 8 at 0x7b48000054c0 by thread T752 (mutexes: write M1):
    #0 std::__1::vector<DB::PageCache::Mmap, std::__1::allocator<DB::PageCache::Mmap>>::begin[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/vector:1429:39 (clickhouse+0x1619c0c4) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 DB::PageCache::getResidentSetSize() const build_docker/./src/Common/PageCache.cpp:205:29 (clickhouse+0x1619c0c4)
    #2 DB::ServerAsynchronousMetrics::updateImpl(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, bool, bool, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, DB::AsynchronousMetricValue, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, DB::AsynchronousMetricValue>>>&) build_docker/./src/Interpreters/ServerAsynchronousMetrics.cpp:84:32 (clickhouse+0x184a121c) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #3 DB::AsynchronousMetrics::update(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, bool) build_docker/./src/Common/AsynchronousMetrics.cpp:1587:5 (clickhouse+0xf27fd72) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #4 DB::AsynchronousMetrics::run() build_docker/./src/Common/AsynchronousMetrics.cpp:362:13 (clickhouse+0xf281838) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #5 DB::AsynchronousMetrics::start()::$_0::operator()() const build_docker/./src/Common/AsynchronousMetrics.cpp:276:62 (clickhouse+0xf28406a) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #6 decltype(std::declval<DB::AsynchronousMetrics::start()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::AsynchronousMetrics::start()::$_0&>(DB::AsynchronousMetrics::start()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf28406a)
    #7 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<DB::AsynchronousMetrics::start()::$_0&, std::__1::tuple<>&>(DB::AsynchronousMetrics::start()::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0xf28406a)
    #8 decltype(auto) std::__1::apply[abi:v15000]<DB::AsynchronousMetrics::start()::$_0&, std::__1::tuple<>&>(DB::AsynchronousMetrics::start()::$_0&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0xf28406a)
    #9 ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:223:13 (clickhouse+0xf28406a)
    #10 decltype(std::declval<DB::AsynchronousMetrics::start()::$_0>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&)::'lambda'()&>(DB::AsynchronousMetrics::start()::$_0&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf28406a)
    #11 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0xf28406a)
    #12 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0xf28406a)
    #13 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0xf28406a)
    #14 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xf205e0e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #15 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xf205e0e)
    #16 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xf205e0e)
    #17 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xf20bcd1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #18 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20bcd1)
    #19 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0xf20bcd1)
    #20 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0xf20bcd1)

  Location is heap block of size 384 at 0x7b4800005400 allocated by main thread:
    #0 operator new(unsigned long, std::align_val_t) <null> (clickhouse+0x71271db) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 void* std::__1::__libcpp_operator_new[abi:v15000]<unsigned long, std::align_val_t>(unsigned long, std::align_val_t) build_docker/./contrib/llvm-project/libcxx/include/new:246:10 (clickhouse+0x1705a29d) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #2 std::__1::__libcpp_allocate[abi:v15000](unsigned long, unsigned long) build_docker/./contrib/llvm-project/libcxx/include/new:267:12 (clickhouse+0x1705a29d)
    #3 std::__1::allocator<std::__1::__shared_ptr_emplace<DB::PageCache, std::__1::allocator<DB::PageCache>>>::allocate[abi:v15000](unsigned long) build_docker/./contrib/llvm-project/libcxx/include/__memory/allocator.h:112:38 (clickhouse+0x1705a29d)
    #4 std::__1::allocator_traits<std::__1::allocator<std::__1::__shared_ptr_emplace<DB::PageCache, std::__1::allocator<DB::PageCache>>>>::allocate[abi:v15000](std::__1::allocator<std::__1::__shared_ptr_emplace<DB::PageCache, std::__1::allocator<DB::PageCache>>>&, unsigned long) build_docker/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:262:20 (clickhouse+0x1705a29d)
    #5 std::__1::__allocation_guard<std::__1::allocator<std::__1::__shared_ptr_emplace<DB::PageCache, std::__1::allocator<DB::PageCache>>>>::__allocation_guard[abi:v15000]<std::__1::allocator<DB::PageCache>>(std::__1::allocator<DB::PageCache>, unsigned long) build_docker/./contrib/llvm-project/libcxx/include/__memory/allocation_guard.h:53:18 (clickhouse+0x1705a29d)
    #6 std::__1::shared_ptr<DB::PageCache> std::__1::allocate_shared[abi:v15000]<DB::PageCache, std::__1::allocator<DB::PageCache>, unsigned long&, unsigned long&, unsigned long&, bool&, bool&, void>(std::__1::allocator<DB::PageCache> const&, unsigned long&, unsigned long&, unsigned long&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:952:48 (clickhouse+0x1705a29d)
    #7 std::__1::shared_ptr<DB::PageCache> std::__1::make_shared[abi:v15000]<DB::PageCache, unsigned long&, unsigned long&, unsigned long&, bool&, bool&, void>(unsigned long&, unsigned long&, unsigned long&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962:12 (clickhouse+0x1705a29d)
    #8 DB::Context::setPageCache(unsigned long, unsigned long, unsigned long, bool, bool) build_docker/./src/Interpreters/Context.cpp:2750:26 (clickhouse+0x1705a29d)
    #9 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/server/Server.cpp:1233:25 (clickhouse+0xf3100c1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #10 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1d92143e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #11 DB::Server::run() build_docker/./programs/server/Server.cpp:416:25 (clickhouse+0xf3014a1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #12 Poco::Util::ServerApplication::run(int, char**) build_docker/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x1d9411d4) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #13 mainEntryClickHouseServer(int, char**) build_docker/./programs/server/Server.cpp:222:20 (clickhouse+0xf2fe603) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #14 main build_docker/./programs/main.cpp:508:21 (clickhouse+0x712920e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)

  Mutex M0 (0x7b4800005470) created at:
    #0 pthread_mutex_lock <null> (clickhouse+0x70a296b) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 std::__1::__libcpp_mutex_lock[abi:v15000](pthread_mutex_t*) build_docker/./contrib/llvm-project/libcxx/include/__threading_support:304:10 (clickhouse+0x20bf50b9) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #2 std::__1::mutex::lock() build_docker/./contrib/llvm-project/libcxx/src/mutex.cpp:38:14 (clickhouse+0x20bf50b9)
    #3 std::__1::unique_lock<std::__1::mutex>::unique_lock[abi:v15000](std::__1::mutex&) build_docker/./contrib/llvm-project/libcxx/include/__mutex_base:122:61 (clickhouse+0x1619bee1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #4 DB::PageCache::getPinnedSize() const build_docker/./src/Common/PageCache.cpp:194:22 (clickhouse+0x1619bee1)
    #5 DB::ServerAsynchronousMetrics::updateImpl(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, bool, bool, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, DB::AsynchronousMetricValue, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, DB::AsynchronousMetricValue>>>&) build_docker/./src/Interpreters/ServerAsynchronousMetrics.cpp:86:60 (clickhouse+0x184a130f) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #6 DB::AsynchronousMetrics::update(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, bool) build_docker/./src/Common/AsynchronousMetrics.cpp:1587:5 (clickhouse+0xf27fd72) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #7 DB::AsynchronousMetrics::start() build_docker/./src/Common/AsynchronousMetrics.cpp:275:5 (clickhouse+0xf27048e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #8 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/server/Server.cpp:1954:23 (clickhouse+0xf3163ba) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #9 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1d92143e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #10 DB::Server::run() build_docker/./programs/server/Server.cpp:416:25 (clickhouse+0xf3014a1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #11 Poco::Util::ServerApplication::run(int, char**) build_docker/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x1d9411d4) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #12 mainEntryClickHouseServer(int, char**) build_docker/./programs/server/Server.cpp:222:20 (clickhouse+0xf2fe603) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #13 main build_docker/./programs/main.cpp:508:21 (clickhouse+0x712920e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)

  Mutex M1 (0x7ffe4749de88) created at:
    #0 pthread_mutex_lock <null> (clickhouse+0x70a296b) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 std::__1::__libcpp_mutex_lock[abi:v15000](pthread_mutex_t*) build_docker/./contrib/llvm-project/libcxx/include/__threading_support:304:10 (clickhouse+0x20bf50b9) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #2 std::__1::mutex::lock() build_docker/./contrib/llvm-project/libcxx/src/mutex.cpp:38:14 (clickhouse+0x20bf50b9)
    #3 std::__1::lock_guard<std::__1::mutex>::lock_guard[abi:v15000](std::__1::mutex&) build_docker/./contrib/llvm-project/libcxx/include/__mutex_base:94:27 (clickhouse+0xf270809) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #4 DB::AsynchronousMetrics::update(std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000l>>>, bool) build_docker/./src/Common/AsynchronousMetrics.cpp:566:21 (clickhouse+0xf270809)
    #5 DB::AsynchronousMetrics::start() build_docker/./src/Common/AsynchronousMetrics.cpp:275:5 (clickhouse+0xf27048e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #6 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/server/Server.cpp:1954:23 (clickhouse+0xf3163ba) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #7 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1d92143e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #8 DB::Server::run() build_docker/./programs/server/Server.cpp:416:25 (clickhouse+0xf3014a1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #9 Poco::Util::ServerApplication::run(int, char**) build_docker/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x1d9411d4) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #10 mainEntryClickHouseServer(int, char**) build_docker/./programs/server/Server.cpp:222:20 (clickhouse+0xf2fe603) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #11 main build_docker/./programs/main.cpp:508:21 (clickhouse+0x712920e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)

  Thread T2046 'VFSRead' (tid=10710, running) created by thread T1228 at:
    #0 pthread_create <null> (clickhouse+0x70a0c6b) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 std::__1::__libcpp_thread_create[abi:v15000](unsigned long*, void* (*)(void*), void*) build_docker/./contrib/llvm-project/libcxx/include/__threading_support:376:10 (clickhouse+0xf20b31e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #2 std::__1::thread::thread<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'(), void>(void&&) build_docker/./contrib/llvm-project/libcxx/include/thread:311:16 (clickhouse+0xf20b31e)
    #3 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool) build_docker/./src/Common/ThreadPool.cpp:217:35 (clickhouse+0xf204b81) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #4 ThreadPoolImpl<std::__1::thread>::scheduleOrThrow(std::__1::function<void ()>, Priority, unsigned long, bool) build_docker/./src/Common/ThreadPool.cpp:288:5 (clickhouse+0xf20d966) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #5 ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./src/Common/ThreadPool.h:203:38 (clickhouse+0xf20d966)
    #6 void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool) build_docker/./src/Common/ThreadPool.cpp:217:35 (clickhouse+0xf2076c3) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #7 ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleOrThrowOnError(std::__1::function<void ()>, Priority) build_docker/./src/Common/ThreadPool.cpp:276:5 (clickhouse+0xf2072ea) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #8 DB::PipelineExecutor::spawnThreads() build_docker/./src/Processors/Executors/PipelineExecutor.cpp:360:15 (clickhouse+0x19f6c831) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #9 DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:306:17 (clickhouse+0x19f6c06e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #10 DB::PipelineExecutor::executeSingleThread(unsigned long) build_docker/./src/Processors/Executors/PipelineExecutor.cpp:239:5 (clickhouse+0x19f6c06e)
    #11 DB::PipelineExecutor::spawnThreads()::$_0::operator()() const build_docker/./src/Processors/Executors/PipelineExecutor.cpp:373:17 (clickhouse+0x19f6cbb0) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #12 decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x19f6cbb0)
    #13 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x19f6cbb0)
    #14 std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x19f6cbb0)
    #15 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::PipelineExecutor::spawnThreads()::$_0, void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x19f6cbb0)
    #16 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xf209b2e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #17 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xf209b2e)
    #18 ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xf209b2e)
    #19 void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xf20defc) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #20 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20defc)
    #21 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0xf20defc)
    #22 decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0xf20defc)
    #23 ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:223:13 (clickhouse+0xf20defc)
    #24 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20de61) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #25 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0xf20de61)
    #26 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0xf20de61)
    #27 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0xf20de61)
    #28 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xf205e0e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #29 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xf205e0e)
    #30 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xf205e0e)
    #31 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xf20bcd1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #32 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf20bcd1)
    #33 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0xf20bcd1)
    #34 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0xf20bcd1)

  Thread T752 'AsyncMetrics' (tid=1529, running) created by main thread at:
    #0 pthread_create <null> (clickhouse+0x70a0c6b) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #1 std::__1::__libcpp_thread_create[abi:v15000](unsigned long*, void* (*)(void*), void*) build_docker/./contrib/llvm-project/libcxx/include/__threading_support:376:10 (clickhouse+0xf20b31e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #2 std::__1::thread::thread<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'(), void>(void&&) build_docker/./contrib/llvm-project/libcxx/include/thread:311:16 (clickhouse+0xf20b31e)
    #3 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool) build_docker/./src/Common/ThreadPool.cpp:217:35 (clickhouse+0xf204b81) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #4 ThreadPoolImpl<std::__1::thread>::scheduleOrThrow(std::__1::function<void ()>, Priority, unsigned long, bool) build_docker/./src/Common/ThreadPool.cpp:288:5 (clickhouse+0xf205036) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #5 ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&) build_docker/./src/Common/ThreadPool.h:203:38 (clickhouse+0xf2705a6) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #6 std::__1::__unique_if<ThreadFromGlobalPoolImpl<true>>::__unique_single std::__1::make_unique[abi:v15000]<ThreadFromGlobalPoolImpl<true>, DB::AsynchronousMetrics::start()::$_0>(DB::AsynchronousMetrics::start()::$_0&&) build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:714:32 (clickhouse+0xf2705a6)
    #7 DB::AsynchronousMetrics::start() build_docker/./src/Common/AsynchronousMetrics.cpp:276:14 (clickhouse+0xf2705a6)
    #8 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/server/Server.cpp:1954:23 (clickhouse+0xf3163ba) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #9 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1d92143e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #10 DB::Server::run() build_docker/./programs/server/Server.cpp:416:25 (clickhouse+0xf3014a1) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #11 Poco::Util::ServerApplication::run(int, char**) build_docker/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x1d9411d4) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #12 mainEntryClickHouseServer(int, char**) build_docker/./programs/server/Server.cpp:222:20 (clickhouse+0xf2fe603) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)
    #13 main build_docker/./programs/main.cpp:508:21 (clickhouse+0x712920e) (BuildId: e5ab52147b76bbe68895173dde1993385f7bf5d2)

SUMMARY: ThreadSanitizer: data race build_docker/./contrib/llvm-project/libcxx/include/__utility/swap.h:36:7 in std::__1::enable_if<is_move_constructible<DB::PageCache::Mmap*>::value && is_move_assignable<DB::PageCache::Mmap*>::value, void>::type std::__1::swap[abi:v15000]<DB::PageCache::Mmap*>(DB::PageCache::Mmap*&, DB::PageCache::Mmap*&)
==================
```

</details>


> Stateless tests (tsan, s3 storage) [1/5] — fail: 1, passed: 1237, skipped: 22 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60875/ba8a2eb10d1a2a2beb0fd72b5191c3f4b2fdbd34/stateless_tests__tsan__s3_storage__[1_5].html)

`02784_parallel_replicas_automatic_decision_join` https://github.com/ClickHouse/ClickHouse/issues/60625
